presto-main/src/main/java/com/facebook/presto/operator/scalar/JoniRegexpFunctions.java

```diff
@@ -211,34 +215,43 @@ public static Block regexpExtractAll(@SqlType("varchar(x)") Slice source, @SqlTy
     public static Block regexpExtractAll(@SqlType("varchar(x)") Slice source, @SqlType(JoniRegexpType.NAME) Regex pattern, @SqlType(StandardTypes.BIGINT) long groupIndex)
     {
         Matcher matcher = pattern.matcher(source.getBytes());
+        int nextStart = 0;
+        int offset = getMatchingOffset(matcher, nextStart, source.length(), false);
+        if (offset == -1) {
+            return EMPTY_BLOCK;
+        }
+
         validateGroup(groupIndex, matcher.getEagerRegion());
-        BlockBuilder blockBuilder = VARCHAR.createBlockBuilder(null, 32);
+        ArrayList<Integer> matches = new ArrayList<>(10);
```

Comment:
Seems like this could be an IntArrayList to avoid boxing overheads?

Reply:
Except I'm getting consistently worse perf numbers like below! Maybe the JIT is doing good job with local box/unbox.
```
sql_regexp_extract_alll ::  809.307 cpu ms ::    0B peak memory :: in   15K,      0B,   18.5K/s,      0B/s :: out     1,  10.7MB,       1/s,  13.2MB/s
Aug 18, 2021 2:31:26 PM com.facebook.airlift.log.Logger info
sql_regexp_replace ::  763.541 cpu ms ::    0B peak memory :: in   15K,      0B,   19.6K/s,      0B/s :: out     1,  21.7MB,       1/s,  28.4MB/s
```

Reply:
The integer boxing cache is indeed a frequent source of strange results when values are in the range of -128,127 (which seems likely to be involved here much more so than some kind of real workloads)- but since there’s nothing like identity hash / equality in play the benchmarks themselves I do wonder how many warmup vs measurement iterations are used and with what JVM flags, etc. I think it’s possible that what you’re measuring is that ArrayList methods are benefiting from getting optimized (C2) compilation whereas the IntArrayList methods are not.

Anyway, this is more of a “I can’t create a reasonable theory of why the boxed pointers would be faster than the primitive arrays” that doesn’t involve hard to introspect details of the JIT code, so feel free to leave it as-is